### ISSUE TITLE
Your app is fatally flawed and not secure!

### ISSUE THREAD

**Issue by vanitasvitae** [YoeriNijs/NoteBuddy#1]
**TL;DR:**
```
private static final String PASSWORD = "jU%C?xy?^jmf8b&d--_XXhKfNvMX_d8tsymBu8jQuBrfe+pxnseDbGKWjVbM7!p5pwaxF9jmDahpUk7qCx@DL^UxBz=k=v!d!e2^";
```

**Now for real:**
The only thing you do is check, whether the hash of the password of the user corresponds to the stored hash before using the static password stored as plaintext in your source code!
Adding salt makes no difference here, since it is also accessible to an attacker. You can "crack" the "encryption" simply by adding a single "!" to this [line](https://github.com/YoeriNijs/NoteBuddy/blob/master/app/src/main/java/nl/yoerinijs/notebuddy/activities/LoginActivity.java#L150), recompiling the app and typing in a wrong password!

The encryption of the notes is totally independent from the password the user chose. All parts of the puzzle used to decrypt the notes are stored as plaintext in the source code or the apps preferences!

Please read about how encryption works and fix these issues ASAP!

I don't want to be mean or something, but you are hurting innocent users that trust your word when reading your apps description.

**Comment by YoeriNijs** [comment]
Thanks for your reply, @vanitasvitae. I was following [the example class by Tozny](https://github.com/tozny/java-aes-crypto/blob/master/aes-crypto-sample-app/app/src/main/java/com/tozny/aes_crypto_sample/MainActivity.java) to implement encryption. They state it is not wrong to generate a key with a provided password. However, according to the articles I have read and due to your feedback, it is needed to not store is somewhere.

First, I was thinking to use a passphrase to generate the key, without saving the password somewhere. However, this could be a potential problem, because I want to check whether the user can login as well. Secondly, I thought it should be possible to use the random generate key function by Tozny and store it somewhere in the KeyStore, but this is quite new for me. Moreover, on the internet, only examples exist with generating a new private/public key pair. Thus, without using a custom generated key. 

I am quite stuck in here, mainly due to the fact this is all quite new for me (this is my first Android app). Do you have an idea to fix these problems fast?

**Comment by vanitasvitae** [comment]
I imagine, if you replace the static key and use instead a key derived from the user password that would drastically improve the encryption. That way you'll lose the option to regain access with a security question though (I never heard of an encryption app with such a feature though).

What you could do is the following:
```
User password -> derive -> keyA -> de/encrypt -> masterKeyA -> de/encrypt -> Notes
User question -> derive -> keyB -> de/encrypt -> masterKeyB -> de/encrypt -> Notes
```
That way you decoupled the user password and question from the masterKey used to encrypt the Notes, while still have reasonable security (given that you create the masterKey once from random data). From the user password you derive keyA, which is used to encrypt/decrypt the masterKey. Same goes for the answer to the security question. masterKeyA and masterKeyB are the same key (the masterKey), but encrypted for/with keyA/keyB.

Thats just a quick sketch how to solve the issue with the static keys. You could take a look at how LUKS/LVM key management works (Linux Harddrive encryption; it features multiple keys for the same data).

Kind regards vanitasvitae

**Comment by YoeriNijs** [comment]
Fixed by adding key derivation. Not sure if I need to generate the master key and if this is the correct way to go, but it should be better than previous versions nevertheless. Link: https://github.com/YoeriNijs/NoteBuddy/tree/7fec1a0312dd22e376f1249cbce7ae336da971c2
